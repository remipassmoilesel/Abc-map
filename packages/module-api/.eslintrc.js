module.exports = {
  extends: ['eslint:recommended', 'plugin:@typescript-eslint/recommended', 'plugin:react/recommended', 'plugin:react-hooks/recommended'],
  settings: {
    react: {
      version: 'detect',
    },
  },
  plugins: ['no-only-tests', '@typescript-eslint'],
  env: {
    node: true,
    mocha: true,
    es6: true,
  },
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: './tsconfig.json',
  },
  rules: {
    'no-only-tests/no-only-tests': 'error',
    'no-console': 1,
    '@typescript-eslint/explicit-module-boundary-types': 0,
    '@typescript-eslint/no-explicit-any': 0,
    '@typescript-eslint/ban-types': 0,
    '@typescript-eslint/explicit-member-accessibility': [
      'error',
      {
        accessibility: 'explicit',
        overrides: {
          constructors: 'no-public',
        },
      },
    ],
    '@typescript-eslint/no-use-before-define': [
      'error',
      {
        functions: false,
        classes: false,
      },
    ],
    'prefer-promise-reject-errors': 'error',
    'no-unused-vars': 'off',
    '@typescript-eslint/no-unused-vars': [
      'error',
      {
        varsIgnorePattern: 'logger',
      },
    ],
    'max-len': ['error', { code: 160 }],
    '@typescript-eslint/no-floating-promises': ['error'],
    'react-hooks/exhaustive-deps': ['error'],
  },
};
