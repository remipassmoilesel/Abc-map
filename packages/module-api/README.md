# Module API

This package exposes typings and helpers that you can use to create Abc-Map modules.

Most of the objects typed here are only available at `runtime` in Abc-Map, not in tests.

To bootstrap a module, see [developer documentation](https://gitlab.com/abc-map/abc-map/-/tree/master/documentation/6_modules.md).
