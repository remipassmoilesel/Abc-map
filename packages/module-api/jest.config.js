/*
 * For a detailed explanation regarding each configuration property, visit:
 * https://jestjs.io/docs/en/configuration.html
 */

module.exports = {
  testEnvironment: 'jsdom',
  coverageDirectory: 'coverage',
  coveragePathIgnorePatterns: ['/node_modules/'],
  coverageProvider: 'babel',
  coverageReporters: ['text', 'html'],
  errorOnDeprecated: true,
  roots: ['src'],
  moduleNameMapper: {
    '\\.css$': '<rootDir>/src/__mocks__/style.js',
    '\\.scss$': '<rootDir>/src/__mocks__/style.js',
  },
  setupFiles: ['<rootDir>/scripts/test-setup.js'],
  transformIgnorePatterns: [],
};
