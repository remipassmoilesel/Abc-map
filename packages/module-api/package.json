{
  "name": "@abc-map/module-api",
  "version": "1.2.0",
  "license": "AGPL-3.0-or-later",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "files": [
    "build",
    "README.md"
  ],
  "scripts": {
    "build": "yarn run build:scss && yarn run build:ts",
    "build:scss": "copy-and-watch 'src/**/*.scss' build",
    "build:ts": "tsc",
    "clean": "rm -rf build",
    "clean-build": "yarn run clean && yarn run build && cp src/**/*.scss build/",
    "dep-check": "depcruise -c .dependency-cruiser.js src",
    "lint": "prettier --loglevel warn --check . && eslint './src/**/{*.tsx,*.ts}' --max-warnings=0",
    "lint-fix": "prettier --loglevel warn --write . && eslint './src/**/{*.tsx,*.ts}' --max-warnings=0 --fix",
    "public-publish": "scripts/public-publish.sh",
    "test": "tsc && CI=true jest src --verbose",
    "test:interactive": "jest src --watch",
    "watch": "concurrently 'yarn run watch:ts' 'yarn run watch:scss'",
    "watch:scss": "copy-and-watch 'src/**/*.scss' build --watch",
    "watch:ts": "tsc --watch | sed -u 's/\\x1Bc//g'"
  },
  "dependencies": {
    "loglevel": "^1.8.1",
    "ol": "7.2.2",
    "react": "^18.2.0",
    "react-hook-form": "^7.41.2",
    "source-map-support": "^0.5.21"
  },
  "devDependencies": {
    "@babel/core": "^7.20.7",
    "@babel/preset-env": "^7.20.2",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@testing-library/react": "^13.4.0",
    "@testing-library/user-event": "^14.4.3",
    "@types/jest": "^29.2.4",
    "@types/react": "^18.0.26",
    "@types/sinon": "^10.0.13",
    "@types/uuid": "^9.0.0",
    "@typescript-eslint/eslint-plugin": "^5.59.1",
    "@typescript-eslint/parser": "^5.59.1",
    "babel-jest": "^29.3.1",
    "concurrently": "^7.6.0",
    "copy-and-watch": "^0.1.6",
    "dependency-cruiser": "^12.1.1",
    "eslint": "^8.39.0",
    "eslint-plugin-no-only-tests": "^3.1.0",
    "eslint-plugin-react": "^7.32.2",
    "eslint-plugin-react-hooks": "^4.6.0",
    "jest": "^29.3.1",
    "jest-environment-jsdom": "^29.3.1",
    "prettier": "^2.8.1",
    "react-dom": "^18.2.0",
    "react-test-renderer": "18.2.0",
    "sinon": "^15.0.1",
    "typescript": "^5.0.4",
    "uuid": "^9.0.0"
  }
}
