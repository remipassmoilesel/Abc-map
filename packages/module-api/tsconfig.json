{
  "include": ["src/**/*.ts", "src/**/*.tsx"],
  "exclude": ["node_modules"],
  "compilerOptions": {
    "outDir": "build",
    "target": "es5",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noFallthroughCasesInSwitch": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "downlevelIteration": true,
    "jsx": "react-jsx",
    "sourceMap": true,
    "declaration": true
  }
}
