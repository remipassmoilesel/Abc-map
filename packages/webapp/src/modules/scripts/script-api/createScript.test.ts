/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { createScript } from './createScript';
import { ScriptContext } from './ScriptContext';
import sinon from 'sinon';
import { ModuleApi } from '@abc-map/module-api';
import { ScriptApi } from './ScriptApi';

describe('createScript', () => {
  it('should work', async () => {
    // Prepare
    const logStub = sinon.stub();
    const context: ScriptContext = {
      moduleApi: { type: 'module-api' } as unknown as ModuleApi,
      scriptApi: { type: 'script-api' } as unknown as ScriptApi,
      log: logStub,
    };

    const content = `\
log(log.toString());
log(JSON.stringify(moduleApi));
log(JSON.stringify(scriptApi));

function wait(n) {
    return new Promise((resolve, reject) => {
        setTimeout(() => resolve(), n);
    });
}

await wait(200);
log('Waited');
`;

    // Act
    const script = createScript(content);
    await script(context);

    // Assert
    expect(script).toMatchSnapshot();
    expect(logStub.args).toMatchSnapshot();
  });
});
