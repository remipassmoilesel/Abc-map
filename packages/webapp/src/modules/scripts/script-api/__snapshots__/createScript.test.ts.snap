// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createScript should work 1`] = `[Function]`;

exports[`createScript should work 2`] = `
Array [
  Array [
    "stub",
  ],
  Array [
    "{\\"type\\":\\"module-api\\"}",
  ],
  Array [
    "{\\"type\\":\\"script-api\\"}",
  ],
  Array [
    "Waited",
  ],
]
`;
